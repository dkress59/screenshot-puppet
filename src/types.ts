import { NextFunction, Request, Response } from 'express'
import { LaunchOptions, PDFOptions, ScreenshotOptions } from 'puppeteer'
export { LaunchOptions, PDFOptions, ScreenshotOptions } from 'puppeteer'

export interface ShotOptions {
	return_url?: string
	method?: 'get' | 'post'
	callback?: CallableFunction
	middleware?: (req: Request, res: Response) => Promise<string | false | null>
	
	/**
	 * 'bin':
	 * – Direct Buffer output
	 * 'json':
	 * – JSON containing JPEG/PDF/PNG Base64 String
	 * 'jpg' | 'pdf' | 'png':
	 * – JPEG/PDF/PNG Base64 String
	 */
	output?: 'bin' | 'jpg' | 'json' | 'pdf' | 'png'
	data?: Record<string, unknown>
	darkMode?: boolean
	override?: boolean | ShotOverrides

	screenshot?: ScreenshotOptions | PDFOptions
	browser?: LaunchOptions
}

export interface ShotOverrides {
	output?: boolean
	data?: boolean
	darkMode?: boolean
	//return_url?: boolean
	width?: boolean
	height?: boolean
}

export interface ShotQuery {
	w?: number
	h?: number
	url?: string
	data?: string // JSON.stringify(Record<string, any>)
	dark?: boolean
	remove?: string // JSON.stringify(List)
	output?: 'bin' | 'jpg' | 'json' | 'pdf' | 'png'
}