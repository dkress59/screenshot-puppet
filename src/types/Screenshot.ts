import { Request } from 'express'
import { PuppetOptions } from './PuppetOptions'

export class Screenshot {

	public w = 1024
	public h = 768
	public url = ''
	public src?: string
	public data?: Record<string, unknown>
	public fileName?: string
	public darkMode = false
	public remove?: string[]
	public errors: unknown[] = []
	public output: 'b64' | 'bin' | 'jpg' | 'json' | 'pdf' | 'png'  = 'json'

	constructor({ path, query }: Request, options?: PuppetOptions) {
		const { w, h, url, data, darkMode, remove, output } = query as Record<string, string>

		this.url = url.substring(0, 7) === 'http://'
			? url
			: 'https://' + url

		if (w) this.w = parseInt(w)
		if (h) this.h = parseInt(h)

		if (remove && typeof remove === 'string' && remove.split(',').length)
			this.remove = remove.split(',')

		const fileExt = path && path !== '/'
			? path.substring(1, path.length - 1).split('.').reverse()[0].toLowerCase()
			: false

		if (fileExt && ['jpg','jpeg','json','pdf','png'].indexOf(fileExt) > -1) {
			this.fileName = path.split('.').splice(-1,1).join('.')}

		if (options?.override === false) {
			if (options.output)
				this.output = options?.output
			if (options.data)
				this.data = options?.data
		} else {
			if (darkMode === 'true')
				this.darkMode = true // ToDo: reassure

			const mergedData = options?.data && data
				? {
					...options.data,
					...JSON.parse(data),
				}
				: data
					? JSON.parse(data)
					: options?.data

			if (mergedData)
				this.data = mergedData

			if (output && ['b64', 'bin', 'jpg', 'json', 'pdf', 'png'].indexOf(output) > -1)
				this.output = output as 'b64' | 'bin' | 'jpg' | 'json' | 'pdf' | 'png'
			else if (fileExt)
				this.output = fileExt === 'jpg'
					? 'jpg'
					: fileExt === 'jpeg'
						? 'jpg'
						: fileExt === 'json'
							? 'json'
							: fileExt === 'pdf'
								? 'pdf'
								: 'png'
		}

	}

}
