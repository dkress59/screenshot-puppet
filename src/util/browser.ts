import util from 'util'
import puppeteer, { Browser } from 'puppeteer'
import { Response } from 'express'
import Screenshot from '../types/Screenshot'
import { logErrorToConsole, logToConsole } from './utils'
// eslint-disable-next-line @typescript-eslint/no-var-requires
const redis = require('redis')
const REDIS = process.env.PUPPET_REDIS || 'redis://127.0.0.1:6379'


export const client = redis.createClient(REDIS)
client.get = util.promisify(client.get)
client.set = util.promisify(client.set)
client.setex = util.promisify(client.setex)
client.exists = util.promisify(client.exists)


export const launchBrowser = async (res?: Response, timeout?: number): Promise<Browser> => {
	process.setMaxListeners(16)
	const browser = await puppeteer
		.launch({
			timeout: timeout ? timeout : 6666,
			defaultViewport: null,
			ignoreHTTPSErrors: true,
			args: ['--no-sandbox', '--disable-setuid-sandbox'], // ToDo: neccessary?
		})
		.catch((e) => {
			if (res)
				res
					.status(500)
					.send({ error: 'error launching puppeteer: ' + e.toString() })
			throw new Error('error launching puppeteer: ' + e.toString())
		})

	return browser
}


export const makeScreenshot = async (browser: Browser, image: Screenshot): Promise<Screenshot> => {
	const { width, height, link, url, darkMode, remove } = image
	if (!link) image.error = 'No link provided for the screen shot.'
	if (!link) return image

	try {

		const page = await browser.newPage()
		if (width && height) await page.setViewport({ width, height })

		if (darkMode)
			await page.emulateMediaFeatures([{
				name: 'prefers-color-scheme',
				value: 'dark',
			}])

		await page.goto(url)

		if (remove)
			remove.map((sel: string) => {
				logToConsole('remove', sel)
				try {
					page.evaluate((sel) => {
						const nodes = document.querySelectorAll(sel)
						if (document.querySelectorAll(sel).length)
							for (let i = 0; i < nodes.length; i++)
								nodes[i].parentNode.removeChild(nodes[i])
					}, sel)
				} catch (error) {
					//image.error = error
					logErrorToConsole(error)
				}
				return
			})

		const screenshot = await page.screenshot()
		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
		// @ts-ignore
		image.source = screenshot.toString('base64') // ToDo: Why isn't btoa() working?

		const cacheId = `${link}-${width}x${height}`
		await client.setex(cacheId, 60 * 60 * 24 * 30, image.source)
		logToConsole(`cache set for ${cacheId}`)

	} catch (error) {

		logToConsole(error)
		image.error = error

	}
	return image
}

export const makePDF = async (doc: string): Promise<false | Buffer> => {
	const path = process.env.LV_REACT_APP_URL + '/invoice/' + doc
	const cookies = [{
		name: 'allowCookies',
		value: 'true',
		path: '/',
		expires: -1,
	}, {
		name: 'ageIsVerified',
		value: 'true',
		path: '/',
		expires: -1,
	}]

	const browser = await launchBrowser()
	const page = await browser.newPage()

	try {

		await page.goto(path, { waitUntil: 'domcontentloaded' })
		await page.setCookie(...cookies)
		await page.reload({ waitUntil: 'networkidle2' })
		await page.emulateMediaType('screen')// ToDo: 'print'

		const file = await page.pdf({ format: 'A4' })
		await client.setex(`lv-pdf/${doc}`, 60 * 60 * 24 * 30, file.toString('base64'))

		await browser.close()
		return file

	} catch (err) {

		logErrorToConsole(err)
		await browser.close()
		return false

	}
}
