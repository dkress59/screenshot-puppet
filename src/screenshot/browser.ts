import { Response } from 'express'
import { Screenshot } from '../util/Screenshot'
import { logErrorToConsole, logToConsole } from '../util/utils'
import puppeteer, { Browser, LaunchOptions, PDFOptions, ScreenshotOptions } from 'puppeteer'


const qualities = new Set([
	0,
	1,
	2,
	3,
	4,
	5,
	6,
	7,
	8,
	9,
	10,
	11,
	12,
	13,
	14,
	15,
	16,
	17,
	18,
	19,
	20,
	21,
	22,
	23,
	24,
	25,
	26,
	27,
	28,
	29,
	30,
	31,
	32,
	33,
	34,
	35,
	36,
	37,
	38,
	39,
	40,
	41,
	42,
	43,
	44,
	45,
	46,
	47,
	48,
	49,
	50,
	51,
	52,
	53,
	54,
	55,
	56,
	57,
	58,
	59,
	60,
	61,
	62,
	63,
	64,
	65,
	66,
	67,
	68,
	69,
	70,
	71,
	72,
	73,
	74,
	75,
	76,
	77,
	78,
	79,
	80,
	81,
	82,
	83,
	84,
	85,
	86,
	87,
	88,
	89,
	90,
	91,
	92,
	93,
	94,
	95,
	96,
	97,
	98,
	99,
	100
])

export const launchBrowser = async (res?: Response, options?: LaunchOptions): Promise<Browser> => {
	process.setMaxListeners(16) // ToDo: options?
	return await puppeteer
		.launch({
			timeout: 6666,
			defaultViewport: null,
			ignoreHTTPSErrors: true,
			args: ['--no-sandbox', '--disable-setuid-sandbox'], // ToDo: neccessary?
			...options,
			headless: true,
		})
		.catch( /* istanbul ignore next */ (error) => {
			if (res)
				res
					.status(500)
					.send({
						message: 'error launching puppeteer',
						// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
						error: error.message,
					})

			throw new Error('error launching puppeteer: ' + JSON.stringify(error)) // breaks jest
		})
}


// eslint-disable-next-line sonarjs/cognitive-complexity
export const makeScreenshot = async (browser: Browser, image: Screenshot, options?: ScreenshotOptions | PDFOptions): Promise<Screenshot> => {

	const { w, h, url, darkMode, remove, output } = image

	const encoding = (output === 'bin')
		? 'binary'
		: 'base64'
	
	const type = options && ('type' in options)
		? options.type
		: output === 'pdf'
			? undefined
			: output === 'jpg'
				? 'jpeg'
				: 'png'

	const quality = (
		options
			&& 'quality' in options
			&& options.quality !== undefined
			&& qualities.has(options.quality)
	)
		? options.quality
		: undefined

	const safeOptions: ScreenshotOptions | PDFOptions = {
		...options,
		encoding,
		path: undefined, // can save to disk
		type,
		quality,
	}

	try {

		const page = await browser.newPage()
		await page.setViewport({ width: w, height: h })

		if (darkMode)
			await page.emulateMediaFeatures([{
				name: 'prefers-color-scheme',
				value: 'dark',
			}])

		await page.goto(url, {
			waitUntil: 'networkidle0'
		})

		if (remove)
			for (const sel of remove) {
				logToConsole('remove', sel)
				try {
					/* istanbul ignore next */
					await page.evaluate((sel) => {
						const nodes = document.querySelectorAll(sel)
						if (document.querySelectorAll(sel).length)
							for (let i = 0; i < nodes.length; i++)
								// eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
								nodes[i].remove()
					}, sel)
				} catch (error) {
					image.errors.push(error)
					logErrorToConsole(error)
				}
			}

		const screenshot = (output === 'pdf')
			? await page.pdf(safeOptions)
			: await page.screenshot(safeOptions)

		image.src = screenshot

	} catch (error) {

		logToConsole(error)
		image.errors.push(error)

	}


	return image

}
