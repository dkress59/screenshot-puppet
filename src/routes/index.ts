import { Request, Response } from 'express'
import { logErrorToConsole, logToConsole } from '../util/utils'
import { launchBrowser, makeScreenshot } from '../util/browser'
import { Screenshot } from '../types/Screenshot'
import { PuppetOptions } from '../types/PuppetOptions'

/**
 * * output â€“ source of truth
 * - default
 * - type
 * - init
 * - request
*/

export const postRouteScreenshot = async (req: Request, res: Response, options?: PuppetOptions): Promise<void> => {

	res.type('application/json')
	const { cached, needed } = req.body
	const browser = await launchBrowser(res)

	const returns = []
	const errors = []
	for await (const image of needed)
		returns.push(
			(async () => {
				const img = new Screenshot(image)
				try {

					const response = await makeScreenshot(browser, img, options?.screenshot)
					if (response.src)
						return response
					else
						errors.push(response)

				} catch (error) {

					img.error = error
					logErrorToConsole(error)
					errors.push(img)

				}
			})()
		)

	Promise.all(returns)
		.then((images) => {
			if (options?.callback) options.callback(images)
			if (images.filter((i) => i && i.error !== undefined).length)
				res.status(200).send(JSON.stringify([...cached, ...images]))
		})
		.catch((err) => {
			logErrorToConsole(err)
			res.status(500).send(JSON.stringify({ error: err }))
		})
		.finally(() => {
			logToConsole('closing browser...')
			browser.close().catch((e: Error) => void e)
		})
}

export const getRouteScreenshot = async (req: Request, res: Response, options?: PuppetOptions): Promise<void> => {
	res.type('application/json')
	
	const image = new Screenshot(req)

	const browser = await launchBrowser(res)

	const response = await makeScreenshot(browser, image, options?.screenshot)

	if (options?.callback) options.callback(response)

	response.src
		? res.status(200).send(JSON.stringify(response))
		: image.error.length
			? res.status(500).send(JSON.stringify(response))
			: res.status(500).send(
				JSON.stringify({
					...image,
					error: 'Error while retreiving screen shot.',
				})
			)

	logToConsole('closing browser...')
	await browser.close()
	return
}