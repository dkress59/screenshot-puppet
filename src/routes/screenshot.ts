import { Response } from 'express'
import Screenshot from '../types/Screenshot'
import { logErrorToConsole, logToConsole } from '../util/utils'
import ParsedQuery from '../types/ParsedQuery'
import { launchBrowser, makeScreenshot } from '../util/browser'

export const postRouteScreenshot = async (req: Request, res: Response): Promise<void> => {

	res.type('application/json')
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	const { cached, needed } = req.body
	const browser = await launchBrowser(res)

	const returns = []
	const errors = []
	for await (const image of needed)
		returns.push(
			(async () => {
				try {

					const response = await makeScreenshot(browser, image)
					if (response.source)
						return response
					else
						errors.push(response)

				} catch (error) {

					console.error(error)
					errors.push({
						...image,
						error,
					})

				}
			})()
		)

	Promise.all(returns)
		.then((images) => {
			if (images.filter((i) => i && i.error !== undefined).length)
				res.status(200).send(JSON.stringify([...cached, ...images]))
		})
		.catch((err) => {
			logErrorToConsole(err)
			res.status(500).send(JSON.stringify({ error: err }))
		})
		.finally(() => {
			logToConsole('closing browser...')
			browser.close().catch((e: Error) => void e)
		})
}

export const getRouteScreenshot = async (req: Request, res: Response): Promise<void> => {
	res.type('application/json')
	
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	const image = new Screenshot(new ParsedQuery(req))
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	image.remove = req.query.remove.split(',')

	const browser = await launchBrowser(res)

	const response = await makeScreenshot(browser, image)
	response.source
		? res.status(200).send(JSON.stringify(response))
		: image.error
			? res.status(500).send(JSON.stringify(response))
			: res.status(500).send(
				JSON.stringify({
					...image,
					error: 'Error while retreiving screen shot.',
				})
			)

	logToConsole('closing browser...')
	await browser.close()
	return
}
