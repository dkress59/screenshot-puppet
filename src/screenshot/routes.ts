import { Request, Response } from 'express'
import { logErrorToConsole, logToConsole } from '../util/utils'
import { launchBrowser, makeScreenshot } from './browser'
import { Screenshot } from '../util/Screenshot'
import { ShotOptions } from '../types'
import queryString from 'query-string'
import { Browser } from 'puppeteer'

const makeOriginURL = (req: Request, options?: ShotOptions) => options?.return_url
	? `${options.return_url}${req.path}${req.params?.filename ?? ''}${
		req.query
			? '?' + queryString.stringify(req.query as Record<string, string>)
			: ''
	}`
	: req.protocol + '://' + req.get('host') + req.originalUrl

export async function getScreenshotRoute(req: Request, res: Response, options?: ShotOptions): Promise<void> {

	const cacheSuccess = options?.middleware && await options.middleware(req, res)
	if (cacheSuccess === false) return

	const image = new Screenshot(req, options)
	if (cacheSuccess) image.src = cacheSuccess

	const browser = cacheSuccess ? {} as Browser : await launchBrowser(res)
	
	const response = cacheSuccess ? image : await makeScreenshot(browser, image, options?.screenshot)
	
	if (options?.callback) options.callback(response)
	
	switch(response.output) {
		case 'bin':
			break
		case 'jpg':
			res.type('jpg')
			break
		case 'pdf':
			res.type('pdf')
			break
		case 'png':
			res.type('png')
			break
		default:
			res.type('json')
			break
	}
	
	const originalUrl = makeOriginURL(req, options)
	
	response.output === 'json'
		? res.status(200).send(JSON.stringify({ response, originalUrl }))
		: response.output === 'bin'
			? res.status(200).send(response.src)
			: res.status(200).send(Buffer.from(response.src as string, 'base64'))
	
	if (browser) {
		browser.close().catch((e: Error) => void e)
		logToConsole(`
			browser closed.
		`)
	}
	return
}

export async function postScreenshotRoute(req: Request, res: Response, options?: ShotOptions): Promise<void> {

	res.type('json')

	if (options?.middleware) await options.middleware(req, res)

	const { cached, needed } = req.body.cached && req.body.needed
		? req.body
		: {
			cached: [],
			needed: req.body
		}
	const browser = needed.length ? await launchBrowser(res) : {} as Browser

	const returns = []
	const errors = []
	for await (const query of needed) {
		const img = new Screenshot({ query, path: req.path } as Request, options)
		try {

			const response = await makeScreenshot(browser, img, options?.screenshot)
			returns.push(response)

		} catch (error) {

			img.errors.push(error)
			logErrorToConsole(error)
			errors.push(img)

		}
	}

	const originalUrl = makeOriginURL(req, options)
	const conditionalErrors = errors.length ? { errors } :  {}

	if (options?.callback) options.callback(returns)
	res.status(200).send(JSON.stringify({
		...conditionalErrors,
		originalUrl,
		response: [...cached, ...returns],
	}))

	if (browser) {
		browser.close().catch((e: Error) => void e)
		logToConsole(`
			browser closed.
		`)
	}
}
