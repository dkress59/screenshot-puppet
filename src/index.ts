import { Request, Response } from 'express'
import { getRouteScreenshot, postRouteScreenshot } from './screenshot/routes'
import { PuppetOptions } from './PuppetOptions'
import { Screenshot as IScreenshot } from './util/Screenshot'

/**
 * * output â€“ source of truth
 * - default
 * - options
 * - ?fileext
 * - ?request
*/

const ScreenshotPuppet = (options?: PuppetOptions | 'get' | 'post'): (req: Request, res: Response) => Promise<void> => {

	if (
		options === 'post'
		|| typeof options === 'object' && 'method' in options && options.method === 'post'
	)
		return async (req: Request, res: Response): Promise<void> =>
			await postRouteScreenshot(req, res, typeof options === 'object'
				? options
				: undefined
			)

	return async (req: Request, res: Response): Promise<void> =>
		await getRouteScreenshot(req, res, typeof options === 'object'
			? options
			: undefined
		)

}

export type Screenshot = IScreenshot

export const getScreenshot = async (req: Request, res: Response): Promise<void> =>
	await getRouteScreenshot(req, res)

export const postScreenshot = async (req: Request, res: Response): Promise<void> =>
	await postRouteScreenshot(req, res)


export default ScreenshotPuppet