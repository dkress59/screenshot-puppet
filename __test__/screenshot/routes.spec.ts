/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Request, Response } from 'express'
import { getScreenshotRoute, postScreenshotRoute } from '../../src/screenshot/routes'
import { ShotOptions } from '../../src/types'

import { Browser, Page, ScreenshotOptions } from 'puppeteer'
import { Screenshot } from '../../src/util/Screenshot'

import { makeScreenshot, launchBrowser } from '../../src/screenshot/browser'
import { mocked } from 'ts-jest/utils'
jest.mock('../../src/screenshot/browser')

const mockedMakeScreenshot = mocked(makeScreenshot, true)
const mockedLaunchBrowser = mocked(launchBrowser, true)

let req: Partial<Request> = {}
const mockResType = jest.fn()
const mockResStatus = jest.fn()
const mockResSend = jest.fn()
const mockResEnd = jest.fn()
let res: Partial<Response> = {
	type: mockResType,
	status: mockResStatus,
	send: mockResSend,
	end: mockResEnd,
}

const env = process.env

describe('Screenshot Routes', () => {

	describe('GET route', () => {
		beforeEach(() => {
			req = {
				path: '/getRoute/',
				query: {
					url: 'https://duckduckgo.com'
				}
			}
			res = {
				type: mockResType.mockReset().mockReturnThis(),
				status: mockResStatus.mockReset().mockReturnThis(),
				send: mockResSend.mockReset().mockReturnThis(),
				end: mockResEnd.mockReset().mockReturnThis(),
			}
			mockedLaunchBrowser.mockImplementation((): Promise<Browser> =>
				new Promise(resolve =>
					resolve({
						close: () => ({
							catch: () => null
						}),
						newPage: () => ({
							setViewport: () => null,
							goto: () => null,
							screenshot: () => 'abc',
							pdf: () => 'abc',
						} as unknown as Page),
					} as unknown as Browser)
				)
			)
			mockedMakeScreenshot.mockImplementation(
				(_browser: Browser, image: Screenshot, _options?: ScreenshotOptions): Promise<Screenshot> =>
					new Promise(resolve => resolve({ ...image, src: 'abc' } as unknown as Screenshot))
			)
		})

		it('returns screenshot', async() => {
			const mockOptions: ShotOptions = {
				return_url: 'http://return.url',
			}
			/* const mockQuery = {}
			const req: Partial<Request> = {
				...req,
				query: {
					url: req.query!.url,
					...mockQuery
				}
			} */

			await getScreenshotRoute(req as Request, res as Response, mockOptions)

			expect(res.type).toHaveBeenCalledWith('json')
			expect(mockedLaunchBrowser).toHaveBeenCalled()
			expect(res.status).toHaveBeenCalledWith(200)
			expect(res.send).toMatchSnapshot()
		})

		it('calls callback', async() => {
			const callback = jest.fn()
			const mockOptions: ShotOptions = {
				return_url: 'http://return.url',
				callback,
			}

			await getScreenshotRoute(req as Request, res as Response, mockOptions)

			expect(callback).toHaveBeenCalled()
		})

		describe('content type', () => {

			it('set json correctly', async() => {
				const mockOptions: ShotOptions = {
					return_url: 'http://return.url',
					// @ts-ignore
					output: 'invalid'
				}

				await getScreenshotRoute(req as Request, res as Response, mockOptions)

				expect(res.type).toHaveBeenCalledWith('json')
			})

			it('set jpeg correctly', async() => {
				const mockOptions: ShotOptions = {
					return_url: 'http://return.url',
					output: 'jpg'
				}

				await getScreenshotRoute(req as Request, res as Response, mockOptions)

				expect(res.type).toHaveBeenCalledWith('jpg')
			})

			it('set pdf correctly', async() => {
				const mockOptions: ShotOptions = {
					return_url: 'http://return.url',
					output: 'pdf'
				}

				await getScreenshotRoute(req as Request, res as Response, mockOptions)

				expect(res.type).toHaveBeenCalledWith('pdf')
			})

			it('set png correctly', async() => {
				const mockOptions: ShotOptions = {
					return_url: 'http://return.url',
					output: 'png',
					override: true,
				}

				await getScreenshotRoute(req as Request, res as Response, mockOptions)

				expect(res.type).toHaveBeenCalledWith('png')
			})

			it('set bin correctly', async() => {
				const mockOptions: ShotOptions = {
					return_url: 'http://return.url',
					output: 'bin'
				}

				await getScreenshotRoute(req as Request, res as Response, mockOptions)

				expect(res.type).not.toHaveBeenCalled()
			})

		})
	
	})

	describe('POST Route', () => {
		beforeEach(() => {
			process.env.NODE_ENV = 'test'
			req = {
				path: '/postRoute/',
				body: [{
					url: 'https://duckduckgo.com'
				}, {
					url: 'https://github.com'
				}],
			}
			res = {
				type: mockResType.mockReset().mockReturnThis(),
				status: mockResStatus.mockReset().mockReturnThis(),
				send: mockResSend.mockReset().mockReturnThis(),
				end: mockResEnd.mockReset().mockReturnThis(),
			}
			mockedLaunchBrowser.mockImplementation((): Promise<Browser> =>
				new Promise((resolve) =>
					resolve({
						close: () => ({
							catch: () => null
						}),
						newPage: () => ({
							setViewport: () => null,
							goto: () => null,
							screenshot: () => 'abc',
							pdf: () => 'abc',
						} as unknown as Page),
					} as unknown as Browser)
				)
			)
		})
		afterAll(() => {
			process.env = env
		})

		it('returns screenshot', async() => {
			const mockOptions: ShotOptions = {
				return_url: 'http://return.url',
			}

			await postScreenshotRoute(req as Request, res as Response, mockOptions)

			expect(res.type).toHaveBeenCalledWith('json')
			expect(mockedLaunchBrowser).toHaveBeenCalled()
			// expect(res.status).toHaveBeenCalledWith(200)
			expect(res.send).toMatchSnapshot()
		})

		it('returns errors', async() => {
			const mockOptions: ShotOptions = {
				return_url: 'http://return.url',
			}

			mockedMakeScreenshot.mockImplementation(() => Promise.reject('rejection'))

			await postScreenshotRoute(req as Request, res as Response, mockOptions)

			expect(res.type).toHaveBeenCalledWith('json')
			expect(res.status).toHaveBeenCalledWith(200)
			expect(res.send).toBeCalledWith('{"errors":[{"w":1024,"h":768,"url":"https://duckduckgo.com","src":"","darkMode":false,"errors":["rejection"],"output":"json"},{"w":1024,"h":768,"url":"https://github.com","src":"","darkMode":false,"errors":["rejection"],"output":"json"}],"originalUrl":"http://return.url/postRoute/","response":[]}')
			expect(res.send).toMatchSnapshot()
		})

		it('calls callback', async() => {
			const callback = jest.fn()
			const mockOptions: ShotOptions = {
				return_url: 'http://return.url',
				callback,
			}

			await postScreenshotRoute(req as Request, res as Response, mockOptions)

			expect(callback).toHaveBeenCalled()
		})
		
	})
	

})