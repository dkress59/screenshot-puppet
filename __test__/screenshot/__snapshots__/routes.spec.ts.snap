// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot Routes GET route returns screenshot 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"response\\":{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://duckduckgo.com\\",\\"src\\":\\"abc\\",\\"darkMode\\":false,\\"errors\\":[],\\"output\\":\\"json\\"},\\"originalUrl\\":\\"http://return.url/getRoute/?url=https%3A%2F%2Fduckduckgo.com\\"}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "end": [MockFunction],
        "send": [MockFunction] {
          "calls": Array [
            Array [
              "{\\"response\\":{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://duckduckgo.com\\",\\"src\\":\\"abc\\",\\"darkMode\\":false,\\"errors\\":[],\\"output\\":\\"json\\"},\\"originalUrl\\":\\"http://return.url/getRoute/?url=https%3A%2F%2Fduckduckgo.com\\"}",
            ],
          ],
          "results": [Circular],
        },
        "status": [MockFunction] {
          "calls": Array [
            Array [
              200,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "type": [MockFunction] {
          "calls": Array [
            Array [
              "json",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`Screenshot Routes POST Route returns errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"errors\\":[{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://duckduckgo.com\\",\\"src\\":\\"\\",\\"darkMode\\":false,\\"errors\\":[\\"rejection\\"],\\"output\\":\\"json\\"},{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://github.com\\",\\"src\\":\\"\\",\\"darkMode\\":false,\\"errors\\":[\\"rejection\\"],\\"output\\":\\"json\\"}],\\"originalUrl\\":\\"http://return.url/postRoute/\\",\\"response\\":[]}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "end": [MockFunction],
        "send": [MockFunction] {
          "calls": Array [
            Array [
              "{\\"errors\\":[{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://duckduckgo.com\\",\\"src\\":\\"\\",\\"darkMode\\":false,\\"errors\\":[\\"rejection\\"],\\"output\\":\\"json\\"},{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://github.com\\",\\"src\\":\\"\\",\\"darkMode\\":false,\\"errors\\":[\\"rejection\\"],\\"output\\":\\"json\\"}],\\"originalUrl\\":\\"http://return.url/postRoute/\\",\\"response\\":[]}",
            ],
          ],
          "results": [Circular],
        },
        "status": [MockFunction] {
          "calls": Array [
            Array [
              200,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "type": [MockFunction] {
          "calls": Array [
            Array [
              "json",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`Screenshot Routes POST Route returns screenshot 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"originalUrl\\":\\"http://return.url/postRoute/\\",\\"response\\":[{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://duckduckgo.com\\",\\"src\\":\\"abc\\",\\"darkMode\\":false,\\"errors\\":[],\\"output\\":\\"json\\"},{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://github.com\\",\\"src\\":\\"abc\\",\\"darkMode\\":false,\\"errors\\":[],\\"output\\":\\"json\\"}]}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "end": [MockFunction],
        "send": [MockFunction] {
          "calls": Array [
            Array [
              "{\\"originalUrl\\":\\"http://return.url/postRoute/\\",\\"response\\":[{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://duckduckgo.com\\",\\"src\\":\\"abc\\",\\"darkMode\\":false,\\"errors\\":[],\\"output\\":\\"json\\"},{\\"w\\":1024,\\"h\\":768,\\"url\\":\\"https://github.com\\",\\"src\\":\\"abc\\",\\"darkMode\\":false,\\"errors\\":[],\\"output\\":\\"json\\"}]}",
            ],
          ],
          "results": [Circular],
        },
        "status": [MockFunction] {
          "calls": Array [
            Array [
              200,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "type": [MockFunction] {
          "calls": Array [
            Array [
              "json",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;
